### PR TITLE
[69][IMP] Check in/Check out attendance entry in auditlog based on tz

### PR DESCRIPTION
**Pull request by YogeshMahera-SerpentCS**
- TASK #[69](https://www.quartile.co/web#id=69&action=771&model=project.task&view_type=form&menu_id=505)
-  Fixed pylint for `sale_order_adj`

### CODE REVIEW

**Comment by YogeshMahera-SerpentCS** [comment]
Hi @TimLai125 can you please review it?

File: auditlog_ext_nrq/models/auditlog_log_line.py [hunk 1]
```diff
@@ -13,21 +14,26 @@ def create(self, vals):
         if vals.get('new_value_text') or vals.get('old_value_text'):
             old_value_text, new_value_text = vals.get(
                 'old_value_text'), vals.get('new_value_text')
+            timezone = pytz.timezone('Japan')
```
Comment (yostashiro) [inline auditlog_ext_nrq/models/auditlog_log_line.py]: @YogeshMahera-SerpentCS I suggested that we take the timezone from the superuser. We do not want to fix Japan in the code.

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: auditlog_ext_nrq/models/auditlog_log_line.py (hunk 1 above)
Comment (TimLai125) [reply auditlog_ext_nrq/models/auditlog_log_line.py]: Sorry @yostashiro, this is my mistake, Yogesh asked me on this yesterday and I told him we should fix the Japan timezone. Sorry for the miscommunication.

Comment (YogeshMahera-SerpentCS) [reply auditlog_ext_nrq/models/auditlog_log_line.py]: Hi, @yostashiro @TimLai125 Amended changes with superuser timezone.:)

File: auditlog_ext_nrq/models/auditlog_log_line.py [hunk 2]
```diff
@@ -13,21 +13,22 @@ def create(self, vals):
         if vals.get('new_value_text') or vals.get('old_value_text'):
             old_value_text, new_value_text = vals.get(
                 'old_value_text'), vals.get('new_value_text')
+            tz = self.env['res.users'].sudo().browse(SUPERUSER_ID).tz
```
Comment (yostashiro) [inline auditlog_ext_nrq/models/auditlog_log_line.py]: Please consider the case where timezone setting is missing, in which case UTC should be applied.

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: auditlog_ext_nrq/models/auditlog_log_line.py (hunk 2 above)
Comment (YogeshMahera-SerpentCS) [reply auditlog_ext_nrq/models/auditlog_log_line.py]: HI @yostashiro Amended Changes

**Review by yostashiro — COMMENTED** [review]
@YogeshMahera-SerpentCS @TimLai125 As I further look at the code, the structure of create() method doesn't look very optimal.  I think it can be further normalized.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -30,7 +30,8 @@ exclude: |
   project_parent/|
   web_export_view/|
   web_responsive/|
-  base_technical_features/
+  base_technical_features/|
+  hr_public_holidays/
```
Comment (TimLai125) [inline .pre-commit-config.yaml]: This is unrelated to this PR, please revert this change.

**Review by TimLai125 — CHANGES_REQUESTED** [review]

**Comment by YogeshMahera-SerpentCS** [comment]
Hi, @TimLai125 optimized code. can you please have look?

File: auditlog_ext_nrq/models/auditlog_log_line.py [hunk 4]
```diff
@@ -1,36 +1,30 @@
...
-                })
-            except Exception:
-                pass
+        tz = self.env['res.users'].sudo().browse(SUPERUSER_ID).tz or 'UTC'
+        if vals.get('old_value_text'):
+            vals.update({'old_value_text': self.convert_time_to_users_timezone(
+                vals.get('old_value_text'), tz)})
+        if vals.get('new_value_text'):
+            vals.update({'new_value_text': self.convert_time_to_users_timezone(
+                vals.get('new_value_text'), tz)})
```
Comment (yostashiro) [inline auditlog_ext_nrq/models/auditlog_log_line.py]:
```suggestion
        for field in ["old_value_text", "new_value_text"]:
            val = vals.get(field)
            if val:
                try:
                    converted_time = fields.Datetime.to_string(
                        fields.Datetime.context_timestamp(
                            self.with_context(tz=tz),
                            fields.Datetime.from_string(val)))                    
                    vals.update({field: converted_time})
                except Exception:
                    pass
```

@YogeshMahera-SerpentCS What I had in mind is something like this.

Comment (YogeshMahera-SerpentCS) [reply auditlog_ext_nrq/models/auditlog_log_line.py]: yes, this is also will good optimize :)

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 5]
```diff
@@ -64,3 +67,44 @@ def test_03_unlink_record(self):
             ('res_id', '=', self.translation_record.id),
         ], limit=1, order='id desc')
         self.assertEqual(log.log_category, 'unlink')
+
+    def test_04_create_vals(self):
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py:69]: @YogeshMahera-SerpentCS It's hard for me to understand the scope of this test, and why the test is done in the way it's written here. Can't we simply feed create() method with specific `old_value_text` and `new_value_text`, and check the converted values against expected results (specific times in Japan tz)?

Comment (YogeshMahera-SerpentCS) [reply auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py:69]: Hi @yostashiro yes right but, Here we do add test-case based on the original flow of audit log module. So if we miss something then it will clarify as an error. Also, you correct directly we can pass the value to create method.

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 6]
```diff
@@ -64,3 +68,51 @@ def test_03_unlink_record(self):
...
+        """
+            Test old value text,new value text with datetime field data
+            Test old value text,new value text without datetime field data
+        """
+        tz = 'Japan'
+        self.test_user = self.env.user.copy()
+        self.env.user.write({'tz': tz})
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]: Why don't we set the timezone to the superuser, following the expected usage?

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 7]
```diff
@@ -64,3 +68,51 @@ def test_03_unlink_record(self):
...
+        old_value_text = datetime.now() - relativedelta(months=1)
+        new_value_text = datetime.now()
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]:
```suggestion
```

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 8]
```diff
@@ -64,3 +68,51 @@ def test_03_unlink_record(self):
...
+        field_id = self.env.ref('base.field_ir_cron_nextcall')
+        audit_log_line = self.env['auditlog.log.line']
+        audit_log_line_01 = \
+            audit_log_line.sudo(self.test_user).create({
+                'field_id': field_id.id,
+                'old_value_text': fields.Datetime.to_string(old_value_text),
+                'new_value_text': fields.Datetime.to_string(new_value_text)
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]:
```suggestion
                "old_value_text": "2020-01-01 01:00:00",
                "new_value_text": "2020-01-02 01:00:00"
```

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 9]
```diff
@@ -64,3 +68,51 @@ def test_03_unlink_record(self):
...
+            })
+        old_value_text = fields.Datetime.context_timestamp(
+            audit_log_line.with_context(tz=tz),
+            old_value_text)
+        old_value_text = fields.Datetime.to_string(old_value_text)
+        new_value_text = fields.Datetime.context_timestamp(
+            audit_log_line.with_context(tz=tz),
+            new_value_text)
+        new_value_text = fields.Datetime.to_string(new_value_text)
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]:
```suggestion
```

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 10]
```diff
@@ -64,3 +68,51 @@ def test_03_unlink_record(self):
...
+
+        # Check the old value data updated based on superuser timezone.
+        self.assertEqual(audit_log_line_01.old_value_text, old_value_text)
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]:
```suggestion
        self.assertEqual(audit_log_line_01.old_value_text, "2020-01-01 10:00:00")
```

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 11]
```diff
@@ -64,3 +68,51 @@ def test_03_unlink_record(self):
...
+
+        # Check the new value data updated based on superuser timezone.
+        self.assertEqual(audit_log_line_01.new_value_text, new_value_text)
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]:
```suggestion
        self.assertEqual(audit_log_line_01.new_value_text, "2020-01-02 10:00:00")
```

**Review by yostashiro — CHANGES_REQUESTED** [review]
@YogeshMahera-SerpentCS Let's not over-complicate the test by adding unnecessary computation of the values.

Please also run pre-commit, and report the situation if it does not work.

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 12]
```diff
@@ -64,3 +68,51 @@ def test_03_unlink_record(self):
...
+        """
+        tz = 'Japan'
+        self.test_user = self.env.user.copy()
+        self.env.user.write({'tz': tz})
+        old_value_text = datetime.now() - relativedelta(months=1)
+        new_value_text = datetime.now()
+        field_id = self.env.ref('base.field_ir_cron_nextcall')
+        audit_log_line = self.env['auditlog.log.line']
+        audit_log_line_01 = \
+            audit_log_line.sudo(self.test_user).create({
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]:
```suggestion
            audit_log_line.sudo().create({
```

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 13]
```diff
@@ -64,3 +68,51 @@ def test_03_unlink_record(self):
...
+        self.assertEqual(audit_log_line_01.old_value_text, old_value_text)
+
+        # Check the new value data updated based on superuser timezone.
+        self.assertEqual(audit_log_line_01.new_value_text, new_value_text)
+
+        old_value_text = 'Test Name'
+        new_value_text = 'Updated Test Name'
+        field_id = self.env.ref('base.field_ir_cron_name')
+        audit_log_line_02 = \
+            audit_log_line.sudo(self.test_user).create({
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]:
```suggestion
            audit_log_line.sudo().create({
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py (hunk 6 above)
Comment (YogeshMahera-SerpentCS) [reply auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]: Yes, the default timezone for the super user is Europe/Brussels so change to Japan.

**Comment by YogeshMahera-SerpentCS** [comment]
Hi, @yostashiro Amended changes , Can you please review it?

File: auditlog_ext_nrq/readme/DESCRIPTION.rst [hunk 14]
```diff
@@ -1,2 +1,2 @@
 This modules does the following:
-  * Allows to user can check the audit log entry based on user timezone only for datetime field.
\ No newline at end of file
+  * Allows to user can check the audit log entry based on user timezone only for datetime field.
```
Comment (yostashiro) [inline auditlog_ext_nrq/readme/DESCRIPTION.rst]:
```suggestion
  * Converts old_value_text and new_value_text to superuser's timezone when the value comes from a datetime field.
```

File: auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py [hunk 15]
```diff
@@ -64,3 +64,45 @@ def test_03_unlink_record(self):
...
+
+    def test_04_create_vals(self):
+        """
+            Test old value text,new value text with datetime field data
+            Test old value text,new value text without datetime field data
+        """
+        tz = 'Japan'
+
+        # Default timezone for super user is Europe/Brussels so change to Japan
+        self.env.user.write({'tz': tz})
```
Comment (yostashiro) [inline auditlog_ext_nrq/tests/test_auditlog_ext_nrq.py]:
```suggestion
        # Japan is 9 hours ahead of UTC.
        self.env["res.users"].browse(SUPERUSER_ID).write({"tz": "Japan"})
```

I think it's better to be explicit that the timezone is being set to the superuser here.

**Review by yostashiro — CHANGES_REQUESTED** [review]
> Please also run pre-commit, and report the situation if it does not work.

@YogeshMahera-SerpentCS Please follow up on this comment as well.

**Comment by YogeshMahera-SerpentCS** [comment]
Hi, @TimLai125 Travis is not triggered in PR. I think the free plan is over? is it something related to this? https://github.com/travis-ci/travis-ci/issues/7643#issuecomment-296367102.
